suite/tests/client-interface/drbbdup-drwrap-test.dll.c

```diff
@@ -0,0 +1,187 @@
+ * DAMAGE.
+ */
+
+/* Tests the drbbdup extension when combined with drrwap but using drwrap
+ * in only a subset of the cases.
+ */
+
+#include "dr_api.h"
+#include "drmgr.h"
+#include "drbbdup.h"
+#include "drwrap.h"
+#include <string.h>
+
+#define CHECK(x, msg)                                                                \
+    do {                                                                             \
+        if (!(x)) {                                                                  \
+            dr_fprintf(STDERR, "CHECK failed %s:%d: %s\n", __FILE__, __LINE__, msg); \
+            dr_abort();                                                              \
+        }                                                                            \
+    } while (0);
+
+/* We assume the app is single-threaded for this test. */
+static uintptr_t case_encoding = 0;
+
+static uintptr_t
+set_up_bb_dups(void *drbbdup_ctx, void *drcontext, void *tag, instrlist_t *bb,
+               bool *enable_dups, bool *enable_dynamic_handling, void *user_data)
+{
+    drbbdup_status_t res;
+
+    CHECK(enable_dups != NULL, "should not be NULL");
+    CHECK(enable_dynamic_handling != NULL, "should not be NULL");
+
+    res = drbbdup_register_case_encoding(drbbdup_ctx, 1);
+    CHECK(res == DRBBDUP_SUCCESS, "failed to register case 1");
+
+    *enable_dups = true;
+    *enable_dynamic_handling = false; /* disable dynamic handling */
+    return 0;                         /* return default case */
+}
+
+static void
+event_analyse_case(void *drcontext, void *tag, instrlist_t *bb, uintptr_t encoding,
+                   void *user_data, void *orig_analysis_data, void **case_analysis_data)
+{
+    if (encoding == 0) {
+        int consec_nop_count = 0;
+        for (instr_t *inst = instrlist_first_app(bb); inst != NULL;
+             inst = instr_get_next_app(inst)) {
+            if (instr_get_opcode(inst) == OP_nop)
+                ++consec_nop_count;
+            else {
+                if (consec_nop_count == 4) {
+                    /* Time to switch modes. */
+                    case_encoding = 1;
+                }
+                consec_nop_count = 0;
+            }
+        }
+    }
+    if (encoding == 1) {
+        /* drwrap guarantees it doesn't need "for_trace", "translating", or
+         * a non-DR_EMIT_DEFAULT return value for drwrap_wrap().
+         */
+        drwrap_invoke_analysis(drcontext, tag, bb, /*for_trace=*/false,
+                               /*translating=*/false, case_analysis_data);
+    }
+}
+
+static void
+event_instrument_instr(void *drcontext, void *tag, instrlist_t *bb, instr_t *instr,
+                       instr_t *where, uintptr_t encoding, void *user_data,
+                       void *orig_analysis_data, void *case_analysis_data)
+{
+    if (encoding == 1) {
+        /* drwrap guarantees it doesn't need "for_trace", "translating", or
+         * a non-DR_EMIT_DEFAULT return value for drwrap_wrap().
```

Comment:
s/drwrap_wrap()/drwrap_invoke_insert()

Reply:
No this means for using wrapping (as opposed to replacing)

Reply:
Now obsolete in any case.